### ISSUE TITLE
Add a way to "drop" a property when it doesn't match the schema

### ISSUE THREAD

**Issue by JaneJeon** [colinhacks/zod#2475]
<details>
<summary> Discussed in https://github.com/colinhacks/zod/discussions/2469 </summary>

<div type='discussions-op-text'>

<sup>Originally posted by **JaneJeon** May 31, 2023</sup>
Hi, I have an object that's coming from outside that I need to put into in some _guaranteed form_, but NOT throw.

Basically, I want to make sure that the object's property matches the validation OR it's not included in the first place.

For a simple example, let's consider an object with a string field `foo`:

```ts
z.object({ foo: z.string().url() }).partial()
```

However, this doesn't work because if `foo` is not a URL, it will throw instead of just dropping the property.

Here's how it would ideally go:

```ts
schema.parse({ foo: 'bar'}) // -> {}
schema.parse({}) // -> {}
schema.parse({ foo: 'https://example.com' }) // -> { foo: 'https://example.com' }
```

Even transformers don't work, because I don't want the `foo` field with an undefined value to be in the object. I just want the key to not even be present in the first place.

So far the only thing that _seems_ like it could work is if I were to create a union of objects (_not_ the `foo` field) in which the `foo` field is present and meets validation, or isn't present at all:

```ts
z.union([ z.object({}).strip(), z.object({ foo: z.string().url() } ])
```

However, I have 5 other properties like `foo`, and that means I'd need a union of 2^5 = 32 schemas, which is just untenable, especially as the number of properties may go up in the future.

Is there any good way to do this other than a MASSIVE union of schemas, i.e. to do this at the field level so that `foo` either is accepted or is dropped (without throwing)?

Thanks.</div>
</details>

Basically, I would like to request for a way to "unset" the key when the value doesn't match the schema. With current methods, there's no way to do this in-schema other than an exponentially increasing combination of schemas.

Thank you.

**Comment by JaneJeon** [comment]
This is still relevant

**Comment by JacobWeisenburger** [comment]
+1

**Comment by JacobWeisenburger** [comment]
Perhaps something like this would help in the meantime.
https://github.com/JacobWeisenburger/zod_utilz#partialsafeparse
